plugins/org.komodo.core/src/org/komodo/repository/search/ObjectSearcher.java

```diff
@@ -0,0 +1,433 @@
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301 USA.
+ */
+package org.komodo.repository.search;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import javax.jcr.Node;
+import javax.jcr.NodeIterator;
+import javax.jcr.Session;
+import javax.jcr.query.Query;
+import javax.jcr.query.QueryManager;
+import javax.jcr.query.QueryResult;
+import org.komodo.repository.ObjectImpl;
+import org.komodo.repository.RepositoryImpl.UnitOfWorkImpl;
+import org.komodo.repository.search.Clause.LogicalOperator;
+import org.komodo.spi.KException;
+import org.komodo.spi.query.sql.SQLConstants;
+import org.komodo.spi.repository.KomodoObject;
+import org.komodo.spi.repository.Repository;
+import org.komodo.spi.repository.Repository.KeywordCriteria;
+import org.komodo.spi.repository.Repository.UnitOfWork;
+import org.komodo.utils.ArgCheck;
+import org.komodo.utils.StringUtils;
+
+/**
+ * Finder class for searching a repository
+ */
+public class ObjectSearcher implements SQLConstants {
+
+    private final Repository repository;
+
+    private Set<FromType> fromTypes;
+
+    private List<Clause> whereClauses;
+
+    private String customWhereClause;
+
+    /**
+     * @param repository the repository to search
+     */
+    public ObjectSearcher(Repository repository) {
+        ArgCheck.isNotNull(repository);
+        this.repository = repository;
+    }
+
+    private Session getSession(final UnitOfWork transaction) {
+        return ((UnitOfWorkImpl)transaction).getSession();
+    }
+
+    private boolean isEmpty(Collection<?> list) {
+        return list == null || list.isEmpty();
+    }
+
+    /**
+     * @return the repository
+     */
+    public Repository getRepository() {
+        return this.repository;
+    }
+
+    /**
+     * Add a node type to the From clause
+     *
+     * @param type the type to be added
+     * @param alias the alias of the type
+     */
+    public void addFromType(String type, String alias) {
+        ArgCheck.isNotEmpty(type);
+        String failMsg = "From clause data should be added prior to where clause data"; //$NON-NLS-1$
+        ArgCheck.isTrue(whereClauses == null, failMsg);
```

Comment:
Can remove checks to allow changing the order and shift the argchecks to the createStatement method.